R/plumber-step.R

```diff
@@ -163,7 +163,15 @@ PlumberEndpoint <- R6Class(
     #' @param req a request object
     #' @return a logical. `TRUE` when endpoint can serve request.
     canServe = function(req){
-      req$REQUEST_METHOD %in% self$verbs && !is.na(stri_match_first_regex(req$PATH_INFO, private$regex$regex)[1,1])
+      if (!is.na(stri_match_first_regex(req$PATH_INFO, private$regex$regex)[1,1])) {
+        if (req$REQUEST_METHOD %in% self$verbs) {
+          return(TRUE)
+        } else {
+          req$methodNotAllowed <- TRUE
+          req$verbsAllowed <- c(req$verbsAllowed, self$verbs)
+        }
+      }
+      return(FALSE)
```

Comment:
I believe the original logic needs to be kept and the new logic needs to be moved inside `default404Handler()`.  There can be a situation where you add handler to `GET /meztez` and add another handler of `POST /meztez`. Following this setup, asking for a `POST /meztez` would return `TRUE` w/ `req$methodNotAllowed == TRUE` as the `GET` route would be processed first.

Inside `default404Handler, the `pr` object should be accessible with `req$pr`.  I have not thought through the `private$regex$regex`

Reply:
methodNotAllowed isTRUE (plumber.R line 697) is not going to be hit if a valid handle is found before. Thus no 405.

But I agree it could be cleaner.

Reply:
> methodNotAllowed isTRUE (plumber.R line 697) is not going to be hit if a valid handle is found before. Thus no 405.

True. Feels odd setting that variable when it's not completely true.

> But I agree it could be cleaner.

Yes. There are too many other things inside the router handler that use this function for me to feel safe. 

---------------

As always, thank you for the help!

Reply:
I'll provide an alternative where we use a separate validPath  function in plumberStep (similar to canServe but do not check verbs).

Reply:
I'm moving back on my words. I would have to replicate `handle` loop with `filt` done in `route` and everything to achieve a similar result inside the `default404handler`.

I'll keep it as it is for the moment, maybe I'll think of something.

Reply:
I don't have a better idea on how to handle it. No rush to merge this one so let it simmer, maybe a better solution will present itself.